### PR TITLE
http2: migrates DATA frame payloads to the visitor API

### PR DESCRIPTION
**Pull request by birenroy**
This slightly simplifies the flow of DATA frame payloads; after this change, the `StreamDataFrameSource` class can be removed.

* `Http2Visitor::OnReadyToSendDataForStream()` is substantially the same implementation as `StreamDataFrameSource::SelectPayloadLength()`.
* `Http2Visitor::SendDataFrame()` is substantially the same as `StreamDataFrameSource::Send()`.

Commit Message: http2: migrates DATA frame payloads to the visitor API
Additional Description:
Risk Level: low
Testing: ran unit and integration tests locally
Docs Changes:
Release Notes: added a note with the runtime feature to disable in case of problems
Runtime guard: `envoy_reloadable_features_http2_use_visitor_for_data`

### CODE REVIEW

File: source/common/http/http2/codec_impl.cc [hunk 1]
```diff
@@ -1692,6 +1699,58 @@ int64_t ConnectionImpl::Http2Visitor::OnReadyToSend(absl::string_view serialized
                              serialized.size());
 }
 
+ConnectionImpl::Http2Visitor::DataFrameHeaderInfo
+ConnectionImpl::Http2Visitor::OnReadyToSendDataForStream(Http2StreamId stream_id,
+                                                         size_t max_length) {
+  StreamImpl* stream = connection_->getStream(stream_id);
+  if (stream == nullptr) {
+    return {-1, false, false};
```
Comment (diannahu) [inline source/common/http/http2/codec_impl.cc:1707]: optional, here and below: Could add `/*name=*/` for the bools.

**Review by diannahu — APPROVED** [review]
Nice, clear mapping for the migration! Not sure if the test results are related, but the change LGTM.

**Comment by RyanTheOptimist** [comment]
Is this a real failure, or a flake?
```
[  FAILED  ] IpVersions/Http2FrameIntegrationTest.AdjustUpstreamSettingsMaxStreams/IPv6_Nghttp2, where GetParam() = 8-byte object <01-00 00-00 00-00 00-00>
```
https://dev.azure.com/cncf/envoy/_build/results?buildId=171029&view=logs&jobId=767be981-567e-57d8-68c3-2140ede0a0bd&j=767be981-567e-57d8-68c3-2140ede0a0bd&t=2181edf2-f610-59f2-c43a-04bb9d0bca00

**Comment by birenroy** [comment]
> Is this a real failure, or a flake?
> 
> ```
> [  FAILED  ] IpVersions/Http2FrameIntegrationTest.AdjustUpstreamSettingsMaxStreams/IPv6_Nghttp2, where GetParam() = 8-byte object <01-00 00-00 00-00 00-00>
> ```
> 
> https://dev.azure.com/cncf/envoy/_build/results?buildId=171029&view=logs&jobId=767be981-567e-57d8-68c3-2140ede0a0bd&j=767be981-567e-57d8-68c3-2140ede0a0bd&t=2181edf2-f610-59f2-c43a-04bb9d0bca00

Test passes with `--runs_per_test=50` when I run it. I think it's a flake, since the behavior being tested does not seem to rely on DATA frames.

File: source/common/http/http2/codec_impl.cc (hunk 1 above)
Comment (birenroy) [reply source/common/http/http2/codec_impl.cc]: Done, thanks!

**Review by diannahu — APPROVED** [review]

**Review by RyanTheOptimist — DISMISSED** [review]

**Comment by RyanTheOptimist** [comment]
Looks like coverage has dropped. Maybe we need to add a tests where the flag is false?

Per-extension coverage failed:
Code coverage for source/common/http is lower than limit of 96.6 (96.4)
Code coverage for source/common/http/http2 is lower than limit of 95.6 (93.8)

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**